### PR TITLE
Remove PROTOTYPEs from QueryBuilders

### PR DESCRIPTION
**Pull request by nik9000**
Now that we've changed queries to be registered by their parsers and readers instead of PROTOTYPEs we can remove the PROTOTYPEs entirely. This does that.

Relates to #17085

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java [hunk 1]
```diff
@@ -34,10 +34,25 @@
 public final class EmptyQueryBuilder extends AbstractQueryBuilder<EmptyQueryBuilder> {
 
     public static final String NAME = "empty_query";
-
-    /** the one and only empty query builder */
     public static final EmptyQueryBuilder PROTOTYPE = new EmptyQueryBuilder();
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java]: Is this still here because its not easy to get rid of or because we are planning to hopefully be able to remove the whole class soon?

File: core/src/main/java/org/elasticsearch/index/query/WildcardQueryBuilder.java [hunk 2]
```diff
@@ -162,7 +178,7 @@ public static WildcardQueryBuilder fromXContent(QueryParseContext parseContext)
         }
 
         if (value == null) {
-            throw new ParsingException(parser.getTokenLocation(), "No value specified for prefix query");
+            throw new ParsingException(parser.getTokenLocation(), "No value specified for wildcard query");
```
Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/index/query/WildcardQueryBuilder.java]: nice ;-)

**Comment by cbuescher** [comment]
Left one minor comment, rest LGTM.

File: core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java [hunk 3]
```diff
@@ -75,6 +76,8 @@ public final String toString() {
             toXContent(builder, EMPTY_PARAMS);
             return builder.string();
         } catch (Exception e) {
+            // So we have a stack trace logged somewhere
+            ESLoggerFactory.getLogger(getClass().getName()).warn("Error building toString", e);;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java]: does not seem harmless, but not sure we should sneak this in here. Also it may be worth discussing what we do when things break here and what we should do instead.

Comment (cbuescher) [inline core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java]: Does or doesn`t? Not quiet sure from your comment if you think this is potentially harmfull or not.

File: core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java [hunk 4]
```diff
@@ -78,6 +77,24 @@ public BoostingQueryBuilder(QueryBuilder positiveQuery, QueryBuilder negativeQue
...
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeQuery(positiveQuery);
+        out.writeQuery(negativeQuery);
+        out.writeFloat(negativeBoost);
+    }
+
+
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/BoostingQueryBuilder.java]: nit: can you remove this additional line break?

File: core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java [hunk 5]
```diff
@@ -97,6 +95,28 @@ public GeoBoundingBoxQueryBuilder(String fieldName) {
...
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeGeoPoint(topLeft);
+        out.writeGeoPoint(bottomRight);
+        type.writeTo(out);
+        validationMethod.writeTo(out);
+    }
+
+
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/GeoBoundingBoxQueryBuilder.java]: one too many line break

File: core/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java [hunk 6]
```diff
@@ -61,6 +59,26 @@ public MatchPhraseQueryBuilder(String fieldName, Object value) {
...
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(fieldName);
+        out.writeGenericValue(value);
+        out.writeVInt(slop);
+        out.writeOptionalString(analyzer);
+    }
+
+
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/MatchPhraseQueryBuilder.java]: here too, sorry :)

File: core/src/main/java/org/elasticsearch/index/query/SpanNearQueryBuilder.java [hunk 7]
```diff
@@ -72,6 +71,26 @@ public SpanNearQueryBuilder(SpanQueryBuilder<?> initialClause, int slop) {
...
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        writeQueries(out, clauses);
+        out.writeVInt(slop);
+        out.writeBoolean(inOrder);
+    }
+
+
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/index/query/SpanNearQueryBuilder.java]: here too

File: core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java (hunk 3 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java]: yea sorry....... should have been "seems harmless". Yet we may want to change what we do here and whether we need logging or not. And not sneak it in among all these other changes.

File: core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/query/EmptyQueryBuilder.java]: I kept this (and one more) because they were used a ton in tests and I figured I should handle that in another commit.

File: core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java (hunk 3 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/support/ToXContentToBytes.java]: Sorry! Yeah, I'm all for moving this to it's own PR.

**Comment by nik9000** [comment]
@javanna fixed the line breaks and removed the toString error reporting so I can make a new PR with it.

**Comment by javanna** [comment]
@nik9000 no need for another review, thanks for addressing

**Comment by nik9000** [comment]
OK! Merging!
